tiledb/sm/query/readers/ordered_dim_label_reader.cc

```diff
@@ -0,0 +1,865 @@
+    }
+
+    // Try the next fragment.
+    f--;
+  }
+}
+
+template <typename LabelType>
+LabelType OrderedDimLabelReader::get_range_as(uint64_t r, uint8_t range_index) {
+  return ranges_[r].typed_data<LabelType>()[range_index];
+}
+
+template <>
+std::string_view OrderedDimLabelReader::get_range_as<std::string_view>(
+    uint64_t r, uint8_t range_index) {
+  if (range_index == 0) {
+    return ranges_[r].start_str();
+  } else {
+    return ranges_[r].end_str();
+  }
+}
+
+template <typename IndexType, typename LabelType, typename Op>
+IndexType OrderedDimLabelReader::search_for_range_fixed(
+    uint64_t r,
+    uint8_t range_index,
+    const IndexType& domain_low,
+    const IndexType& tile_extent) {
+  // Get the value we are looking for.
+  LabelType value = get_range_as<LabelType>(r, range_index);
+
+  // Minimum index to look into.
+  auto non_empty_domain =
+      static_cast<const IndexType*>(non_empty_domain_.data());
+  auto t_min = per_range_array_tile_indexes_[r].min(range_index);
+  auto left = std::max(
+      index_dim_->tile_coord_low(t_min, domain_low, tile_extent),
+      non_empty_domain[0]);
+
+  // Maximum index to look into.
+  auto t_max = per_range_array_tile_indexes_[r].max(range_index);
+  auto right = std::min(
+      index_dim_->tile_coord_high(t_max, domain_low, tile_extent),
+      non_empty_domain[1]);
+
+  // Run a binary search.
+  Op cmp;
+  while (left != right - 1) {
+    // Check against mid.
+    IndexType mid = left + (right - left) / 2;
+    if (cmp(get_value_at<IndexType, LabelType>(mid, domain_low, tile_extent),
+            value)) {
+      right = mid;
+    } else {
+      left = mid;
+    }
+  }
+
+  // Do one last comparison to decide to return left or right.
+  auto bound = range_index == 0 ? left : right;
+  return (cmp(
+             get_value_at<IndexType, LabelType>(bound, domain_low, tile_extent),
+             value)) ?
+             left :
+             right;
+}
+
+template <typename IndexType, typename LabelType>
+void OrderedDimLabelReader::compute_and_copy_range_indexes(
+    IndexType* dest, uint64_t r) {
+  // For easy reference.
+  auto tile_extent = index_dim_->tile_extent().rvalue_as<IndexType>();
+  const IndexType* dim_dom =
+      static_cast<const IndexType*>(index_dim_->domain().data());
+
+  // Set the results.
+  if (increasing_labels_) {
+    dest[0] = search_for_range_fixed<
+        IndexType,
+        LabelType,
+        std::greater_equal<LabelType>>(r, 0, dim_dom[0], tile_extent);
+    dest[1] =
+        search_for_range_fixed<IndexType, LabelType, std::greater<LabelType>>(
+            r, 1, dim_dom[0], tile_extent);
+  } else {
+    dest[0] = search_for_range_fixed<
+        IndexType,
+        LabelType,
+        std::less_equal<LabelType>>(r, 0, dim_dom[0], tile_extent);
+    dest[1] =
+        search_for_range_fixed<IndexType, LabelType, std::less<LabelType>>(
+            r, 1, dim_dom[0], tile_extent);
+  }
+}
```

Comment:
We need an extra check here to verify `dest[0] <= dest[1]`. This occurs when we have a valid range inside a domain that does not contain any values. Ultimately, this should be a valid query that returns no data, but I'm okay with it just throwing an error in the meantime.

Example using test case fixture `CPPOrderedDimLabelReaderFixedDoubleFx`:
```
write_labels(1, 4, {1.0, 2.0, 3.0, 4.0});
std::vector<double> ranges{2.1, 2.8};
auto index = read_labels(ranges);
CHECK(index.size() == 0));
```

Reply:
A related case can occur when querying outside of the non-empty domain. For example, I would expect the following to return zero data:
```
write_labels(1, 4, {1.0, 2.0, 3.0, 4.0});
std::vector<double> ranges{-2.0, 0.0, 5.0, 6.0};
auto index = read_labels(ranges);
CHECK(index.size() == 0));
```